darwin/future/core/types/common.py

```diff
@@ -0,0 +1,27 @@
+from typing import Dict
+
+from pydantic import ConstrainedStr
+
+from darwin.future.data_objects import validators as darwin_validators
+
+
+class TeamSlug(ConstrainedStr):
+    """Team slug type"""
+
+    min_length = 1
+    max_length = 256
+
+    validator = darwin_validators.parse_name
```

Comment:
I actually didn't know you could extend the pydantic subtypes like this, but I don't think this is working the way you intend it to? I can't find docs on extending the class on pydantic page but drilling down into ConstrainedStr in pydantic lib, it's got no `validator` function.

It does have a classmethod `validate` and an internal function `__get_validators__` which yield a bunch of different functions in a chain, including `validate`. 

I could be totally wrong though as I haven't played with extending pydantics types and it could be doing other things under the hood.

Reply:
https://docs.pydantic.dev/latest/usage/types/#custom-data-types

Yeah, see this page on extending str.

Reply:
Yeah, so to be fair, maybe this isn't the perfect way to do this, I'm kind of on the fence tbh.  It's one I've used before, and it works - I haven't tested this one yet - but I've done the same in the past.

The `__get_validators__` dunder yields all the validators one at a time, and then adds the `self.validate` method to the yield list.

This overwrites this, with your validator, which only has to return `True` or `False` to validate the thing.

One possibility is to have an alternate like this:

```python
@classmethod
def validate(self) -> bool:
    self.super().validate and darwin_validators.parse_name
```

If it's too hacky, what would you do instead?  I can assign a normal `constr`, but not with a validator, and you've written one.

Reply:
Have committed a new version, using an init dunder to achieve the same on creation.